src/libs/sys-util-rs/src/numa.rs

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) 2021 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::collections::HashMap;
+use std::fs::DirEntry;
+use std::io::Read;
+use std::path::Path;
+
+use lazy_static::lazy_static;
+
+use crate::cpu::split_cpus;
+use crate::sl;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Invalid CPU number {0}")]
+    InvalidCpu(u32),
+    #[error("Invalid node file name {0}")]
+    InvalidNodeFileName(String),
+    #[error("Can not read from file {0}, {1:?}")]
+    ReadFile(String, #[source] std::io::Error),
+    #[error("Can not open from file {0}, {1:?}")]
+    OpenFile(String, #[source] std::io::Error),
+    #[error("Can not parse CPU info, {0:?}")]
+    ParseCpuInfo(#[from] crate::cpu::Error),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+// global config in UT
+#[cfg(test)]
+lazy_static! {
+    static ref SYS_FS_PREFIX: String = format!(
+        "{}/test/texture",
+        std::env::var("CARGO_MANIFEST_DIR").unwrap()
+    );
+}
+
+// global config in release
+#[cfg(not(test))]
+lazy_static! {
+    static ref SYS_FS_PREFIX: String = String::new();
+}
+
+lazy_static! {
+    // numa node file for UT, we can mock data
+    static ref NUMA_NODE_PATH: String = format!("{}/sys/devices/system/node", &*SYS_FS_PREFIX);
+    // sysfs directory for CPU devices
+    static ref NUMA_CPU_PATH: String = format!("{}/sys/devices/system/cpu", &*SYS_FS_PREFIX);
+}
+
+const NUMA_NODE_PREFIX: &str = "node";
+const NUMA_NODE_CPU_LIST_NAME: &str = "cpulist";
+
+/// Get numa node id for a CPU
+pub fn get_node_id(cpu: u32) -> Result<u32> {
+    let cpu_path = format!("{}/cpu{}", &*NUMA_CPU_PATH, cpu);
+    let path = Path::new(&cpu_path);
+
+    if let Ok(dirs) = path.read_dir() {
+        for d in dirs.flatten() {
```

Comment:
I can't see any reason for the `flatten`.

Reply:
`path.read_dir().iter()` returns `io::Result<DirEntry>`, so flatten is used to unwrap the Result.

Reply:
Ah, I see.  However this means that errors will be ignored - is that what you intend?

Reply:
That's intended. `path.read_dir().iter()` may return errors if there’s some sort of intermittent IO error during iteration. But for proc files, intermittent IO error should not happen.

Reply:
Right... IO failures shouldn't happen on proc files.  So if it does, something is horribly wrong and ignoring it and carrying on is probably a bad idea, right?